### DISCUSSION TITLE
Unified solution for ImageGenerator plugins' command preferences

### DISCUSSION THREAD

**Discussion by introt** [zim-desktop-wiki/zim-desktop-wiki#1613] [Development]
### Background

All of the ImageGenerator plugins shipped with Zim, listed below, include this comment:

    # TODO put these commands in preferences

* diagrameditor.py
* ditaaeditor.py
* equationeditor.py
* gnuplot_ploteditor.py
* gnu_r_ploteditor.py
* scoreeditor.py
* sequencediagrameditor.py

Having hard-coded commands make these plugins fragile and prevent non-developer users from tuning them, be that for the sake of adding some boilerplate or [using a differently-named executable](https://github.com/zim-desktop-wiki/zim-desktop-wiki/discussions/1609#discussioncomment-1373690)

Instead of developing over a handful of solutions to this problem, this discussion thread's aim is to find the best way to solve this once and for all.

**Comment by introt** [comment]
### UI/UX

This part _should_ be simple, but I haven't seen any plugin preferences with a text field, so, uh, _someone needs to look into that_.

Imo a good UX would be a drop-down list with a "Custom..." option at the bottom. The list of other options should be automatically generated from available defaults a la Insert Screenshot.

![image](https://user-images.githubusercontent.com/26069787/134506002-e774f5dc-1e28-446a-9171-e354bd81f6ed.png)

**Comment by jaap-karssenberg** [comment]
One refactoring that will make it easier: today the plugin defines the
plugin_preferences and the preferences dialog turns those into a set of
widget. My intention is to move the creation of the widgets to the plugin
base class, such that the dialog only needs to ask for the widget and show
it, but doesn't know the logic inside. Now it becomes easier to have a
custom widget for this selection in the imagegenerator base class.

Otherwise the selection is limited to widgets supported by the InputForm
class.

Alternative is to extend the InputForm class with an "application" type
input, but I rather not put all kind of specialized stuff in this base
class.

Probable there is commonality with the "CustomizeOpenWithDialog" input in
zim/gui/applications.py -- which is btw also used in the preferences when
setting text editor and browser preferences.

You could even use a custom mimetype and use normal mimetype administration
to save the preference. E.g. "x-zim-imagegenerator/equation". If that makes
re-use of the application logic easier.

**Reply by introt** [reply]
Thanks for your input (pun intended)! Moving the creation of the widgets to the plugin base class sounds like the most flexible solution moving forward.

Regarding the problem at hand, `CustomizeOpenWithDialog` seems like a great fit. Users could store/move/share their custom commands as `.desktop` files.

**Comment by introt** [comment]
### API

* A subclass of `ExtensionBase` which implements a method to check a list of command alternatives defined similarly to Insert Screenshot[1].
  * The user-provided command should be checked before accepting the preference change.
* todo...

0. https://github.com/zim-desktop-wiki/zim-desktop-wiki/blob/develop/zim/plugins/screenshot.py

**Comment by Self-Perfection** [comment]
It also would be nice to get ability to generate images in formats other than PNG. I'd like to store to as many as possible of images in my notes in SVG because:
* SVG images do not become blurry on zooming
* SVG output is usually smaller then corresponding PNG
* SVG is way better suited for VCS: gives somewhat meaningful diff instead of "binary files differ", storing new versions adds much less data to VCS repo (only diff)

SVG output is widespread: ditaa, gnuplot, GraphViz and seqdiag can produce SVG output.

**Reply by introt** [reply]
I'm all for SVG support; would also like to extend ImageGenerator plugins to support wrapping graphical programs like Inkscape and Xournal++ which can also export to png/svg via command line.
